### PR TITLE
import AR wallet tmp

### PR DESCRIPTION
**Pull request by Kamill90**

### CODE REVIEW

File: src/screens/ImportWalletChooseTypeScreen.tsx [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+import { ScreenTemplate, Text, Header, Button, RadioGroup, RadioButton } from 'app/components';
+import { Route, Wallet } from 'app/consts';
+import { HDSegwitP2SHWallet, HDSegwitP2SHArWallet, HDSegwitP2SHAirWallet, BlueApp } from 'app/legacy';
+import { AppSettingsState } from 'app/state/appSettings/reducer';
+import { WalletsActionType } from 'app/state/wallets/actions';
+import { palette, typography } from 'app/styles';
+
+const i18n = require('../../loc');
+
+const WalletTypes = [HDSegwitP2SHArWallet.type, HDSegwitP2SHAirWallet.type, HDSegwitP2SHWallet.type];
```
Comment (KedziaPawel) [inline src/screens/ImportWalletChooseTypeScreen.tsx:14]: Legacy stands for all of the old wallets not only for HDSegwitP2SHWallet

File: src/screens/ImportWalletScreen.tsx [hunk 2]
```diff
@@ -100,7 +113,109 @@ export const ImportWalletScreen = (props: Props) => {
...
+        showSuccessImportMessageScreen();
+      } else {
+        showErrorMessageScreen();
+      }
+    } catch (error) {
+      showErrorMessageScreen();
+    }
+  };
+
+  const saveAIRWallet = async (publicKey: string, recoveryKey: string, mnemonic: string = text) => {
```
Comment (KedziaPawel) [inline src/screens/ImportWalletScreen.tsx:136]: publicKey -> instantPublicKey

File: src/screens/ImportWalletScreen.tsx [hunk 3]
```diff
@@ -32,12 +33,14 @@ interface Props {
       StackNavigationProp<MainCardStackNavigatorParams, Route.ImportWallet>
     >
   >;
+  route: RouteProp<MainCardStackNavigatorParams, Route.ImportWallet>;
   loadWallets: () => Promise<WalletsActionType>;
 }
 
 export const ImportWalletScreen = (props: Props) => {
```
Comment (KedziaPawel) [inline src/screens/ImportWalletScreen.tsx:40]: this component has got so many of its own methods that it shouldn't be a functional component as it creates functions at every rerender. Please make it standard Component

File: src/screens/ImportWalletQRCodeScreen.tsx [hunk 4]
```diff
@@ -118,6 +129,14 @@ class ImportWalletQRCodeScreen extends React.Component<Props, State> {
       // make sure it re-renders
       await sleep(100);
       try {
+        const wallet = new HDSegwitP2SHArWallet();
```
Comment (KedziaPawel) [inline src/screens/ImportWalletQRCodeScreen.tsx:132]: so have do we add Air wallet in this flow? how can we import legacy wallet if we are always navigating for scanning no matter what type of wallet we want import?

File: src/screens/ImportWalletQRCodeScreen.tsx [hunk 5]
```diff
@@ -118,6 +129,14 @@ class ImportWalletQRCodeScreen extends React.Component<Props, State> {
...
+
+        wallet.setMnemonic(text);
```
Comment (KedziaPawel) [inline src/screens/ImportWalletQRCodeScreen.tsx:134]: setMnemonic has got mnemonic validation that you need to handle

File: src/screens/ImportWalletScreen.tsx (hunk 3 above)
Comment (KedziaPawel) [inline src/screens/ImportWalletScreen.tsx:40]: same as below, it should be normal Component

File: src/screens/ImportWalletScreen.tsx [hunk 6]
```diff
@@ -100,7 +113,109 @@ export const ImportWalletScreen = (props: Props) => {
...
+      }
+    } catch (error) {
+      showErrorMessageScreen();
+    }
+  };
+
+  const saveAIRWallet = async (publicKey: string, recoveryKey: string, mnemonic: string = text) => {
+    try {
+      const wallet = new HDSegwitP2SHAirWallet();
+      wallet.setMnemonic(mnemonic);
```
Comment (KedziaPawel) [inline src/screens/ImportWalletScreen.tsx:139]: setMnemonic and addPublicKey are prepared to throw errors if invalid, you should use them at every step of importing wallets so the users gets instant feedback if mnemonic or public key is wrong. Check how this works at wallet creation

File: src/screens/ImportWalletScreen.tsx [hunk 7]
```diff
@@ -100,7 +113,109 @@ export const ImportWalletScreen = (props: Props) => {
     }
   };
 
+  const saveARWallet = async (pubKeyHex: string, mnemonic: string = text) => {
+    try {
+      const wallet = new HDSegwitP2SHArWallet();
+      wallet.setMnemonic(mnemonic);
```
Comment (KedziaPawel) [inline src/screens/ImportWalletScreen.tsx:119]: same as below, validate at every step

File: src/screens/ImportWalletScreen.tsx [hunk 8]
```diff
@@ -19,88 +18,228 @@ import {
...
+        this.showErrorMessageScreen();
+      }
+    } catch (error) {
+      this.showErrorMessageScreen();
+    }
+  };
+
+  saveAIRWallet = async (wallet: HDSegwitP2SHAirWallet, instantPublicKey: string, recoveryKey: string) => {
+    try {
+      wallet.addPublicKey(instantPublicKey);
```
Comment (KedziaPawel) [inline src/screens/ImportWalletScreen.tsx:180]: still, you don't validate every step of scanning, you have to validate recoveryPublicKey at first if it is not valid throw and error if valid navigate the user to scan instant public key

File: src/screens/ImportWalletQRCodeScreen.tsx [hunk 9]
```diff
@@ -1,313 +0,0 @@
-import { CompositeNavigationProp } from '@react-navigation/native';
```
Comment (KedziaPawel) [inline src/screens/ImportWalletQRCodeScreen.tsx:1]: appreciation for merging the logic :)

File: src/screens/ImportWalletScreen.tsx [hunk 10]
```diff
@@ -19,88 +18,228 @@ import {
...
   };
 
-  const onChangeText = (mnemonic: string) => {
-    setText(mnemonic);
-    setValidationError('');
-    if (isButtonDisabled !== (mnemonic.length === 0)) {
-      setIsButtonDisabled(!isButtonDisabled);
+  onChangeText = (mnemonic: string) => {
+    this.setState({ text: mnemonic, validationError: '' });
+    if (this.state.isButtonDisabled !== (mnemonic.length === 0)) {
```
Comment (KedziaPawel) [inline src/screens/ImportWalletScreen.tsx:86]: I would remove the isButtonDisabled variable and make function like canSubmit() => {
const { text, validationError} = this.state
return !validationError && !text.trim();
}
this is more flexible solution

File: src/screens/ImportWalletScreen.tsx [hunk 11]
```diff
@@ -19,88 +18,228 @@ import {
...
 
-  const onImportButtonPress = async () => {
+  onImportButtonPress = async () => {
     Keyboard.dismiss();
     CreateMessage({
       title: i18n.message.creatingWallet,
       description: i18n.message.creatingWalletDescription,
       type: MessageType.processingState,
-      asyncTask: () => importMnemonic(text),
+      asyncTask: () => this.importMnemonic(this.state.text),
```
Comment (KedziaPawel) [inline src/screens/ImportWalletScreen.tsx:80]: trim the text

File: src/screens/ImportWalletScreen.tsx [hunk 12]
```diff
@@ -19,88 +18,228 @@ import {
...
     });
   };
 
-  const onChangeText = (mnemonic: string) => {
-    setText(mnemonic);
-    setValidationError('');
-    if (isButtonDisabled !== (mnemonic.length === 0)) {
-      setIsButtonDisabled(!isButtonDisabled);
+  onChangeText = (mnemonic: string) => {
+    this.setState({ text: mnemonic, validationError: '' });
```
Comment (KedziaPawel) [inline src/screens/ImportWalletScreen.tsx:85]: why do we reset validationError here on changing mnemonic? this error is for the name field as I see

File: src/screens/ImportWalletChooseTypeScreen.tsx [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+  };
+
+  onSelect = (selectedIndex: number) =>
+    this.setState({
+      selectedIndex,
+    });
+
+  get validationError(): string | undefined {
+    if (this.state.isLoading) return;
+    const walletLabels = BlueApp.getWallets().map((wallet: Wallet) => wallet.label) || [];
```
Comment (KedziaPawel) [inline src/screens/ImportWalletChooseTypeScreen.tsx:47]: you wrote a helper for this

File: src/screens/ImportWalletScreen.tsx [hunk 14]
```diff
@@ -19,88 +19,226 @@ import {
...
+    this.setState({ text: mnemonic });
   };
 
-  const onScanQrCodeButtonPress = () => {
-    props.navigation.navigate(Route.ImportWalletQRCode);
+  onLabelChange = (value: string) => {
+    const validationError = isWalletLableInUse(value) && i18n.wallets.importWallet.walletInUseValidationError;
+    this.setState({
+      label: value,
+      validationError,
```
Comment (KedziaPawel) [inline src/screens/ImportWalletScreen.tsx:91]: here you are setting validationError as boolean if isWalletLableInUse(value) would give `false` and it is declared as string in State interface. Assign empty string if isWalletLableInUse(value) would give false

File: src/screens/ImportWalletScreen.tsx [hunk 15]
```diff
@@ -19,88 +19,226 @@ import {
...
+      this.props.navigation.navigate(Route.IntegrateKey, {
+        onBarCodeScan: instantPublicKey => {
+          try {
+            wallet.addPublicKey(instantPublicKey);
+            this.addRecoveryKey(wallet);
+          } catch (_) {
+            this.showErrorMessageScreen();
+          }
+        },
+        title: i18n.wallets.publicKey.recoverySubtitle,
```
Comment (KedziaPawel) [inline src/screens/ImportWalletScreen.tsx:155]: instant subtitle and description i guess

File: src/screens/ImportWalletScreen.tsx [hunk 16]
```diff
@@ -19,88 +19,226 @@ import {
...
+        });
+      },
+      withLink: false,
+      title: i18n.wallets.publicKey.recoverySubtitle,
+      description: i18n.wallets.publicKey.recoveryDescription,
+    });
+  };
+
+  importMnemonic = async (mnemonic: string) => {
+    mnemonic.trim();
```
Comment (KedziaPawel) [inline src/screens/ImportWalletScreen.tsx:234]: trim() doesn't change the original string, it returns a new one, so this operation does nothing here. You have to assign it to new value like `const mnemonicTrimmed = mnemonic.trim()`

File: src/screens/ImportWalletScreen.tsx [hunk 17]
```diff
@@ -218,34 +356,42 @@ export const ImportWalletScreen = (props: Props) => {
...
-            title={i18n.wallets.importWallet.scanQrCode}
-            onPress={onScanQrCodeButtonPress}
+  render() {
+    const { validationError, text } = this.state;
+    return (
+      <ScreenTemplate
+        footer={
+          <>
+            <Button
+              disabled={!text || !!validationError}
```
Comment (KedziaPawel) [inline src/screens/ImportWalletScreen.tsx:366]: label is also needed here probably, `!text || !!validationError || !label`

Comment (Kamill90) [reply src/screens/ImportWalletScreen.tsx:366]: If label is not provided wallet will be named as other imported wallet with its type as name

File: src/screens/ImportWalletScreen.tsx [hunk 18]
```diff
@@ -19,150 +19,289 @@ import {
...
+            this.addRecoveryKey(wallet);
+          } catch (_) {
+            this.showErrorMessageScreen();
+          }
+        },
+        title: i18n.wallets.importWallet.scanWalletAddress,
+        description: i18n.wallets.importWallet.scanWalletAddressDescription,
+        withLink: false,
+      });
+    } catch (_) {}
```
Comment (KedziaPawel) [inline src/screens/ImportWalletScreen.tsx:159]: display error

File: src/screens/ImportWalletScreen.tsx [hunk 19]
```diff
@@ -19,150 +19,289 @@ import {
...
+        });
+      },
+      withLink: false,
+      title: i18n.wallets.importWallet.scanPublicKey,
+      description: i18n.wallets.importWallet.scanPublicKeyDescription,
+    });
+  };
+
+  importMnemonic = async (mnemonic: string) => {
+    const trimmedMemonic = memonic.trim();
```
Comment (KedziaPawel) [inline src/screens/ImportWalletScreen.tsx:234]: typo mnemonic, app would crash

**Pull request merged**